From 7675543d1b0340ff9689b7e3fdd5b323dfe18edf Mon Sep 17 00:00:00 2001
From: Yulei Zhang <yulei.zhang@intel.com>
Date: Fri, 29 Jan 2016 11:28:24 +0800
Subject: [PATCH 337/403] vgt: Remove shadow ppgtt creation from shadow
 context creation

Change the shadow ppgtt creation policy, it won't be created in
shadow context creation notification, instead it will be created
when the context first submitted. As after win10 introduce the
WDDM2.0 driver model, the root pointer attached in context when
it is created may not be the one associated with the context
when submitted to execute.
---
 drivers/gpu/drm/i915/vgt/execlists.c |    2 --
 drivers/gpu/drm/i915/vgt/gtt.c       |   20 ++++++++++++--------
 2 files changed, 12 insertions(+), 10 deletions(-)

diff --git a/drivers/gpu/drm/i915/vgt/execlists.c b/drivers/gpu/drm/i915/vgt/execlists.c
index 274cabe..e8a8e1c 100644
--- a/drivers/gpu/drm/i915/vgt/execlists.c
+++ b/drivers/gpu/drm/i915/vgt/execlists.c
@@ -1217,8 +1217,6 @@ static struct execlist_context *vgt_create_execlist_context(
 		}
 	}
 
-	vgt_el_create_shadow_ppgtt(vgt, ring_id, el_ctx);
-
 	trace_ctx_lifecycle(vgt->vm_id, ring_id,
 			el_ctx->guest_context.lrca, "create");
 	return el_ctx;
diff --git a/drivers/gpu/drm/i915/vgt/gtt.c b/drivers/gpu/drm/i915/vgt/gtt.c
index f018d39..6e3d431 100644
--- a/drivers/gpu/drm/i915/vgt/gtt.c
+++ b/drivers/gpu/drm/i915/vgt/gtt.c
@@ -2494,18 +2494,22 @@ bool ppgtt_update_shadow_ppgtt_for_ctx(struct vgt_device *vgt,
 	/* compare the page table root pointer stored in guest context and
 	 * in shadow page table, update the mapping if it is not aligned
 	 */
-	for (i = 0; i < el_ctx->ppgtt_mm->page_table_entry_cnt; ++ i) {
-		ppgtt_get_rootp_from_ctx(g_state, &ctx_rootp, i);
-		ppgtt_get_guest_root_entry(mm, &pt_ctx_rootp, i);
+	if (mm) {
+		for (i = 0; i < el_ctx->ppgtt_mm->page_table_entry_cnt; ++ i) {
+			ppgtt_get_rootp_from_ctx(g_state, &ctx_rootp, i);
+			ppgtt_get_guest_root_entry(mm, &pt_ctx_rootp, i);
 
-		if (ctx_rootp.val64 != pt_ctx_rootp.val64)
-			break;
-	}
+			if (ctx_rootp.val64 != pt_ctx_rootp.val64)
+				break;
+		}
 
-	if (i != el_ctx->ppgtt_mm->page_table_entry_cnt) {
+		if (i != el_ctx->ppgtt_mm->page_table_entry_cnt) {
+			if (vgt_el_create_shadow_ppgtt(vgt, el_ctx->ring_id, el_ctx))
+				rc = false;
+		}
+	} else {
 		if (vgt_el_create_shadow_ppgtt(vgt, el_ctx->ring_id, el_ctx))
 			rc = false;
 	}
-
 	return rc;
 }
-- 
1.7.10.4

